src/txacme/client.py

```diff
@@ -771,6 +783,17 @@ def _send_request(self, method, url, *args, **kwargs):
                             b'content-type', [None])[0])))
                 .addActionFinish())
 
+    def stop(self):
+        """
+        Stops the operation.
+
+        This cancels pending operations and does cleanup.
+
+        :return: When operation is done.
+        :rtype: Deferred[None]
+        """
+        return self._agent._pool.closeCachedConnections()
```

Comment:
We should file a ticket in Twisted about getting public API for this.

Reply:
well, the Twisted API let you create a pool, and the you pass it to the agent and then you can do whatever you want with that pool :)

So maybe change txacme to initialized the Twisted API with a pool... I know that is extra work.

I am always initializing the agent with a custom pool as I want to know when TCP connections are made and closed.

At the same time, I am using txacme with my own agent..

So in terms of API, I would rather have a method on the Agent API itself to signal that the agent can be shut down..

Below is an example of the API that I use:

```
class PersistentAgent(Agent, AgentMixin):
    """
    HTTP/HTTPS Client agent with a connection pool.

    When `credentials` are set it will make the requests using
    HTTP Basic authentication.
    """

    def __init__(
        self, event_emitter, pool_size=3, ssl_context=None, credentials=None,
            ):
        pool = AuditedHTTPConnectionPool(
            self._reactor, event_emitter=event_emitter, persistent=True)
        pool.retryAutomatically = False
        pool.maxPersistentPerHost = pool_size
        self.emitEvent = event_emitter
        super(PersistentAgent, self).__init__(
            reactor=self._reactor,
            contextFactory=self._getContextFactory(ssl_context),
            connectTimeout=CONNECT_TIMEOUT,
            pool=pool,
            )
        self._setAuthorization(credentials)

    def closePersistentConnections(self):
        """
        Close persistent connection associated with this agent.
        """
        return self._pool.closeCachedConnections()

```

Reply:
I don't know how to fix this in a clean way.

In theory, someone can pass a non-persistent agent to txacme, and then this code will break.

One option would be to have JWSClient.__init__(onStop=callback) and then users can pass a callback to do the cleanup for their custom pool.

I don't like this option, as I think that the JWSClient client should have its own stop method which will also take care 
of cancelling the current pending request.

I tried to update the code so that the pending request is also cancelled...but it assumes that only one request is active at the same time